asserts/snapasserts/validation_sets.go

```diff
@@ -27,20 +27,67 @@ import (
 
 	"github.com/snapcore/snapd/asserts"
 	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/snap/naming"
 )
 
+// InstalledSnap holds the minimal details about an installed snap required to
+// check the validation set.
+type InstalledSnap struct {
+	naming.SnapRef
+	Revision snap.Revision
+}
+
+// NewInstalledSnap creates InstalledSnap.
+func NewInstalledSnap(name, snapID string, revision snap.Revision) *InstalledSnap {
+	return &InstalledSnap{
+		SnapRef:  naming.NewSnapRef(name, snapID),
+		Revision: revision,
+	}
+}
+
+func (e *ValidationSetsConflictError) Error() string {
+	buf := bytes.NewBufferString("validation sets are in conflict:")
+	for _, err := range e.Snaps {
+		fmt.Fprintf(buf, "\n- %v", err)
+	}
+	return buf.String()
+}
+
+// ValidationSetsValidationError describes an error arising
+// from validation of snaps against ValidationSets.
+type ValidationSetsValidationError struct {
+	// snapName -> validationSetKey
+	MissingSnaps       map[string][]string
+	InvalidSnaps       map[string][]string
+	WrongRevisionSnaps map[string][]string
+	Sets               map[string]*asserts.ValidationSet
+}
+
 // ValidationSetsConflictError describes an error where multiple
 // validation sets are in conflict about snaps.
 type ValidationSetsConflictError struct {
 	Sets  map[string]*asserts.ValidationSet
 	Snaps map[string]error
 }
 
-func (e *ValidationSetsConflictError) Error() string {
-	buf := bytes.NewBufferString("validation sets are in conflict:")
-	for _, err := range e.Snaps {
-		fmt.Fprintf(buf, "\n- %v", err)
+func (e *ValidationSetsValidationError) Error() string {
+	buf := bytes.NewBufferString("validation sets assertions are not met:")
+	printDetails := func(header string, details map[string][]string) {
+		if len(details) == 0 {
+			return
+		}
+		fmt.Fprintf(buf, "\n%s:", header)
+		for snapName, validationSetKeys := range details {
+			fmt.Fprintf(buf, "\n- %s", snapName)
+			for _, key := range validationSetKeys {
+				fmt.Fprintf(buf, "\n  - validation set: %s", key)
```

Comment:
I suspect we'll usually have not that many validation sets in play so this could go in a parenthesis on the same line as the snap name?

Reply:
Good idea, done.